client/src/main/java/org/eclipse/hono/client/ConnectionLifecycle.java

```diff
@@ -0,0 +1,71 @@
+/**
+ * Copyright (c) 2019 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+
+package org.eclipse.hono.client;
+
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+
+/**
+ * Provides access to the life cycle events of a connection to a Hono service.
+ *
+ */
+public interface ConnectionLifecycle {
+
+    /**
+     * Adds a handler to be notified when the connection is lost unexpectedly.
+     * 
+     * @param handler The handler.
+     */
+    void addDisconnectHandler(Handler<Void> handler);
```

Comment:
I am not so fond of those generic `Handler<Void>` types, Couldn't we add our own ones, providing a proper interfaces. And mapping this internally to `Handler<Void>`?

Reply:
You mean an interface like 
```
interface DisconnectHandler {
  void onDisconnect();
}
```
?

Reply:
Yes. I would prefer that over `Handler<Void>`. Alternatively you could also use something more generic like `VoidHandler` (or some other name). My primary concern here is the useless `Void` object type.